[Master] Transaction / Invoice Associations

Bob-IT opened the pull request:
I have been tinkering with this on and off and adding here for comment.
The main changes are...
The first few commits involve moving the existing transaction association dialog to a neutral named dialog-assoc file and moving the gnc-split-reg functions to this new file.
A couple of enhancement bugs for sorting and reloading the transaction list.
A new dialog has been added for when the path head is changed which will update the existing relative paths based on the user request to make them absolute ones.
Have changed from using a file chooser widget to a button so that they should get a native file dialog.
Added the use of glyphs for the association cell if there current font supports it.
The next few commits deal with adding the ability to add an association to the business items Invoice/Bill and Voucher and reusing the existing association dialogs. I have used a 'link button' on the plugin page below the notes field which is displayed only when there is an association. 
@gjanssens  Not sure if this is the best location but it is a start point.

@christopherlam I need help with the commit 'Bug 796531' which you commented on and added some scheme code. To fix the issue I have created a new anchor for the association so when clicked the association will open out side of the webkit window just like the existing association code. With the change to trep-engine.scm, tests 90 and 93 fail but other changes are needed, column title, display option, maybe follow 'enable links', what ever you think.

christopherlam commented on gnucash/report/trep-engine.scm:
```diff
@@ -1226,6 +1226,16 @@ be excluded from periodic reporting.")
+                                      "text-cell"
+                                      (gnc:html-transaction-association-anchor (xaccSplitGetParent split) (_ "Association")))
+                                       " ")))))
+
```
That's rather impressive. Here's a follow-up commit that you can fixup to this one. Minor nitpick: Wouldn't `trans-association-guid` be more consistent?

```
modified   gnucash/report/trep-engine.scm
@@ -937,6 +937,7 @@ be excluded from periodic reporting.")
       (list (N_ "Other Account Code")           "j"  (_ "Display the other account code?") #f)
       (list (N_ "Shares")                       "k"  (_ "Display the number of shares?") #f)
       (list (N_ "Price")                        "l"  (_ "Display the shares price?") #f)
+      (list (N_ "Association")                  "l5" (_ "Display the transaction association") #f)
       ;; note the "Amount" multichoice option in between here
       (list optname-grid                        "m5" (_ "Display a subtotal summary table. This requires Display/Amount being 'single") #f)
       (list (N_ "Running Balance")              "n"  (_ "Display a running balance?") #f)
@@ -1061,6 +1062,7 @@ be excluded from periodic reporting.")
                      (opt-val gnc:pagename-display (N_ "Other Account Name"))))
           (cons 'shares (opt-val gnc:pagename-display (N_ "Shares")))
           (cons 'price (opt-val gnc:pagename-display (N_ "Price")))
+          (cons 'association (opt-val gnc:pagename-display "Association"))
           (cons 'amount-single (eq? amount-setting 'single))
           (cons 'amount-double (eq? amount-setting 'double))
           (cons 'common-currency (opt-val gnc:pagename-general optname-common-currency))
@@ -1226,15 +1228,17 @@ be excluded from periodic reporting.")
                                   "number-cell"
                                   (xaccSplitGetAmount split)))))
 
-               (add-if 'association-column?
+               (add-if (column-uses? 'association)
                        (vector ""
                                (lambda (split transaction-row?)
-                                 (let ((url (xaccTransGetAssociation (xaccSplitGetParent split))))
-                                   (if (not (string-null? url))
-                                     (gnc:make-html-table-cell/markup
-                                      "text-cell"
-                                      (gnc:html-transaction-association-anchor (xaccSplitGetParent split) (_ "Association")))
-                                       " ")))))
+                                 (let ((url (xaccTransGetAssociation
+                                             (xaccSplitGetParent split))))
+                                   (and (not (string-null? url))
+                                        (gnc:make-html-table-cell/markup
+                                         "text-cell"
+                                         (gnc:html-transaction-association-anchor
+                                          (xaccSplitGetParent split)
+                                          (_ "📎"))))))))
 
                (add-if (column-uses? 'price)
                        (vector (_ "Price")
```

Bob-IT replied:
@christopherlam Thanks for the update and the suggested change which I have done.

christopherlam commented on gnucash/report/trep-engine.scm:
```diff
@@ -1226,6 +1228,20 @@ be excluded from periodic reporting.")
+                                          (gnc:html-transaction-association-anchor
+                                           (xaccSplitGetParent split)
+                                           (_ "📎"))
+                                           (_ "📎"))))))))
```
ahem.

Bob-IT replied:
Not sure what you mean by this?

christopherlam replied:
There should be only one `(_ "📎"))` -- delete line 1242

christopherlam replied:
Ah sorry didn't see your `opt-use-links?`... my bad. The indenting made it confusing.

christopherlam replied:
This is why emacs indent *strictly* according to parentheses level as follows:
```scheme
               (add-if (column-uses? 'association)
                       (vector ""
                               (lambda (split transaction-row?)
                                 (let ((url (xaccTransGetAssociation
                                             (xaccSplitGetParent split))))
                                   (and (not (string-null? url))
                                        (gnc:make-html-table-cell/markup
                                         "text-cell"
                                         (if opt-use-links?
                                             (gnc:html-transaction-association-anchor
                                              (xaccSplitGetParent split)
                                              (_ "📎"))
                                             (_ "📎"))))))))
```

Bob-IT replied:
There was me just about to point that out, what should the indenting be and I will change it.

christopherlam replied:
Thinking back... if `opt-use-links?` is #FALSE then we shouldn't even display the paperclip :laughing:

Bob-IT replied:
OK, I will change that and push later.

Bob-IT replied:
It does show that there is an association even if they can not click on it. If the user does not want it to show like that, they just do not display the column.

gjanssens commented on gnucash/register/register-gnome/gnucash-sheet-private.c:
```diff
@@ -479,6 +479,20 @@ draw_cell (GnucashSheet *sheet,
+                text = GLYPH_PAPERCLIP;
+            else
+                text = GLYPH_LINK;
+        }
```
Can `text` ever be NULL in this function or is it always one of " ", "f" or "l" ?

If it's never null it's likely to be more efficient to test `*text != ' '` and `*text == 'f'` rather than going through g_strcmp0. If it can be NULL you may have to guard the whole test with a `if (text)`

gjanssens replied:
A bigger issue here is though that I don't think this is the spot to do this. You have added code in the generic sheet that now needs to distinguish based on the presence of a very specific column. The sheet is supposed to be column type agnostic.
Instead you should define a new cell type for the association that does this glyph replacement for you and returns either a single letter or a glyph as text when requested for it.

Bob-IT commented:
@gjanssens I have created an assoc cell as suggested and made appropriate changes to be used, I did wonder if the changes to AsscoCell should be added to RecnCell but then it is probably good to keep them separate.
Do not know why travis is failing on an unrelated test, 109 test-report-utilities, maybe it is my github clone as it fails on my master but not Gnucash.
Will await feedback.

christopherlam commented:
This is bug 797139.

jralls commented:
@Bob-IT Bug 797139 is fixed so if you'll rebase and force-push Travis should pass.

jralls commented on gnucash/register/register-core/assoccell.h:
```diff
@@ -0,0 +1,99 @@
+#include "basiccell.h"
+
+#define GLYPH_PAPERCLIP "\360\237\223\216"
+#define GLYPH_LINK      "\360\237\224\227"
```
Please comment these with their unicode code points:
```
#define GLYPH_PAPERCLIP "\360\237\223\216"  // Codepoint  U+1F4CE
#define GLYPH_LINK      "\360\237\224\227"         // Codepoint U+1F517
```
These should be put somewhere that the sheet and report code can get to them too so there's only one definition.

Bob-IT replied:
Done.

jralls commented on gnucash/register/register-gnome/gnucash-sheet.c:
```diff
@@ -789,6 +789,29 @@ gnucash_sheet_finalize (GObject *object)
+#define GLYPH_LINK      "\360\237\224\227"
+
+gboolean
+gnucash_sheet_use_glyphs (GnucashSheet *sheet)
```
Can't this code be made a function of assoccell instead of sheet so that the determination of whether to use them is colocated with the code that does use them. Speaking of which, I don't see any code that actually enables using them.

Bob-IT replied:
I did not think it could but having tried, yes it can so moved.

jralls commented on gnucash/report/trep-engine.scm:
```diff
@@ -1226,6 +1228,20 @@ be excluded from periodic reporting.")
+                                              (xaccSplitGetParent split)
+                                              (_ "📎"))
+                                             (_ "📎"))))))))
+
```
Did you mean to display the same character for both legs of the if? That aside, does it make sense to put the emoji-set characters in the Scheme code? Aside from it being better to have only one place where the codepoints are set so that it's easy to change them if we decide to, you can't send them like that to the html code, they need to be the corresponding html enitities. If they're going to be translated here they should be translated in the register too, and they need a translator comment explaining what they're for.

Bob-IT replied:
Yes, I have now changed them to be the same as the register column title.

jralls commented on libgnucash/app-utils/gnc-ui-util.c:
```diff
@@ -903,6 +903,56 @@ gnc_get_reconcile_flag_order (void)
+ * gnc_get_association_valid_flags                                  *
+ *   return a string containing the list of associated flags        *
+ *                                                                  *
+ * Returns: the i18n'd reconciled flags string                      *
```
They're not internationalized.

Bob-IT replied:
Copy and paste error, changed

bhardwajs commented on libgnucash/app-utils/gnc-ui-util.h:
```diff
@@ -189,6 +189,13 @@ const char * gnc_get_reconcile_str (char reconciled_flag);
 const char * gnc_get_reconcile_valid_flags (void);
 const char * gnc_get_reconcile_flag_order (void);
 
+#define WASSOC 'w'
```
Can this be `constexpr char WASSOC = 'w'`? I see the use in `gnc_get_association_str` and `gnc_get_association_flag_order` where `char` would fit in directly.

gjanssens replied:
`constexpr` is a C++11 feature. This is C code. So no, I don't think this can be used here. It's a valid improvement in a future C++ conversion though.

gjanssens commented on gnucash/gnome-utils/dialog-assoc-utils.c:
```diff
@@ -0,0 +1,454 @@
+        else if (g_str_has_prefix (uri,"file:") && !g_str_has_prefix (uri,"file://"))
+            part = uri + strlen ("file:");
+
+        if (part)
```
If I read this properly `part` will NULL if prefix is "file://"

Bob-IT replied:
I will push my changes at the weekend. I have slightly changed this and removed the 'if (part)' and also the second parameter to the interior if statements as not required.
Basically the test is for 'file:' and 'file:/' and if found remove that part so we can store a relative path which does not start with a '/'. Paths starting 'file://' are treated as absolute paths which obviously includes 'file:///' format.

gjanssens replied:
> If I read this properly `part` will NULL if prefix is "file://"

Nevermind. I'm clearly still jetlagged :(

gjanssens commented on gnucash/gnome-utils/dialog-assoc-utils.c:
```diff
@@ -0,0 +1,454 @@
+            return g_strdup (part);
+        }
+    }
+    return g_strdup (uri);
```
No. This processing is for internal storage. Absolute paths are stored as file uris, relative parts as just simple relative paths. The idea is to store in uri format as much as possible so remote uris and system local uris can be treated as much as possible in the same way. However there's not valid format for relative paths, so those are stored as just that. This makes it easier to combine the relative path with a base uri later on as well.
In previous implementations relative paths were stored as file:<relative path> or file:/<relative path>. But those are non-standard and add extra processing overhead later on so in earlier discussions we decided to move away from those.

gjanssens commented on gnucash/gnome-utils/dialog-assoc-utils.c:
```diff
@@ -0,0 +1,454 @@
+        if (g_str_has_prefix (uri,"file:/") && !g_str_has_prefix (uri,"file://"))
+            part = uri + strlen ("file:/");
+        else if (g_str_has_prefix (uri,"file:") && !g_str_has_prefix (uri,"file://"))
+            part = uri + strlen ("file:");
```
IIRC I wrote the original gnc_uri_utils under the false assumption uri's would start with file:// instead of file:///. So current code other than the file associations also work under this assumption. I'm fine with cleaning that up, though that's probably best done in a separate PR.

Bob-IT replied:
I have changed this, the test is for 'file:' and 'file:/' with that part being removed from the string so that relative paths are stored without a scheme start and without a '/' if relative. Absolute uri paths do store with 'file:///'

Bob-IT commented on gnucash/gnome-utils/dialog-assoc-utils.c:
```diff
@@ -0,0 +1,454 @@
+    gchar *file_path = NULL;
+    gchar *ret_value = NULL;
+
+    if (path_head && g_strcmp0 (path_head, "") != 0) // not default entry
```
Changed, I tend to just use what I know, will try to remember.

Bob-IT commented on gnucash/gnome-utils/dialog-assoc-utils.h:
```diff
@@ -0,0 +1,38 @@
+#ifndef DIALOG_ASSOC_UTILS_H
+#define DIALOG_ASSOC_UTILS_H
+
+gchar * gnc_assoc_get_unescape_uri (const gchar *path_head, const gchar *uri, gchar *uri_scheme);
```
Done.

Bob-IT commented on gnucash/gnome-utils/dialog-assoc-utils.c:
```diff
@@ -0,0 +1,454 @@
+     */
+    trans_hash = g_hash_table_new (g_direct_hash, g_direct_equal);
+
+    /* Get list of Accounts */
```
Extracted transactions and invoice to separate functions.

Bob-IT commented on gnucash/gnome/dialog-assoc.h:
```diff
@@ -20,9 +20,14 @@
+void gnc_assoc_open_uri (GtkWindow *parent, const gchar *uri);
+
+void gnc_assoc_business_dialog (GtkWindow *parent);
+void gnc_assoc_trans_dialog (GtkWindow *parent);
```
Done.

Bob-IT commented on gnucash/gnome/dialog-assoc.c:
```diff
@@ -0,0 +1,1102 @@
+            g_free (run_scheme);
+        }
+        else
+            gnc_error_dialog (GTK_WINDOW (parent), "%s", _("This associated entry has no valid scheme."));
```
After some more testing it is not required so removed.

Bob-IT commented on gnucash/gnome/dialog-assoc.c:
```diff
@@ -0,0 +1,1102 @@
+    const gchar   *uri;
+
+    if (!gtk_tree_model_get_iter (model, &iter, path))
+        return; /* path describes a non-existing row - should not happen */
```
Changed.

Bob-IT commented on gnucash/gnome-utils/dialog-assoc-utils.c:
```diff
@@ -0,0 +1,454 @@
+{
+    gboolean path_head_set = FALSE;
+    gchar *file_path = NULL;
+    gchar *ret_value = NULL;
```
I do try to do that, this one got missed.

Bob-IT commented on gnucash/gnome/dialog-assoc.c:
```diff
@@ -0,0 +1,1102 @@
+                break;
+            default:
+//            case GNC_INVOICE_CUST_INVOICE:
+//            case GNC_INVOICE_CUST_CREDIT_NOTE:
```
Changed.

Bob-IT commented on gnucash/gnome/dialog-assoc.c:
```diff
@@ -0,0 +1,1102 @@
+        xaccQueryAddSingleAccountMatch (query, acc, QOF_QUERY_AND);
+
+        /* Run the query */
+        for (splits = qof_query_run (query); splits; splits = splits->next)
```
Had not used qof collections in the past so never occurred to me to use, changed.

Bob-IT commented on gnucash/gnome/dialog-assoc.c:
```diff
@@ -0,0 +1,1102 @@
+ *                                                                  *
+ * Args:   parent  - the parent of the window to be created         *
+ * Return: nothing                                                  *
+\********************************************************************/
```
Done.

jralls replied:
Umm, it's still here, and since it starts with `/**` Doxygen will pick it up.

Bob-IT commented:
I have made the recommended changes and some others I found, bit disappointed that you found so many g_free's after checking several times, need to be more careful/observant.

Bob-IT commented on gnucash/gnome/dialog-assoc.c:
```diff
@@ -0,0 +1,1102 @@
+
+        /* This should never be true, but be paranoid */
+        if (!invoice)
+            return;
```
Changed.

Bob-IT commented on gnucash/gnome/dialog-assoc.c:
```diff
@@ -0,0 +1,1102 @@
+            g_free (dialog_uri);
+        }
+    }
+    if (result == GTK_RESPONSE_REJECT)
```
Changed.

Bob-IT commented on gnucash/gnome/dialog-assoc.c:
```diff
@@ -0,0 +1,1102 @@
+/***********************************************************************/
+
+static gboolean
+assoc_set_path_head_label (GtkWidget *path_head_label, const gchar *path_head)
```
Changed.

Bob-IT commented:
OK I have gone through this again and I think I have sorted all the g_free errors, created a new cell type assoc which handles all the glyph parts for the assoc cell. Split some functions and changed the way the list of associations is collected. It would be nice to get this in 4.0.

jralls commented on gnucash/register/register-core/assoccell.h:
```diff
@@ -0,0 +1,101 @@
+void gnc_assoc_cell_set_string_getter (AssocCell *cell,
+                                       AssocCellStringGetter getter);
+
+/** note that chars is copied into the AssocCell directly, but remains
+ * the "property" of the caller.  The caller must maintain the chars
+ * pointer, and the caller must setup a mechanism to 'free' the chars
+ * pointer.  The rationale is that you may have many AssocCell objects
+ * that use the same set of flags -- this saves you an alloc/free for
+ * each cell.  - warlord  2001-11-28
+ */
```
I think `chars` should be `flags`. The note should be preceded by `@param flags` to make it completely clear what it refers to, and the signature line is unnecessary.

jralls commented:
I couldn't find any missed leaks and the code generally looks OK. There are still a ton of comments in the implementation files that are going to get fed to Doxygen because the ornate frames begin with `/**`, but are otherwise not in Doxygen format. That doesn't really affect the code, just our sanity later, so I've marked it approved.

Bob-IT commented:
@gjanssens Do you have any concerns, would like to push this.
